pycontrails/models/apcemm/adapter.py

```diff
@@ -0,0 +1,831 @@
+
+    @overload
+    def eval(self, source: None = ..., **params: Any) -> NoReturn: ...
+
+    def eval(self, source: Flight | None = None, **params: Any) -> Flight | NoReturn:
+        """Run APCEMM simulation on flight.
+
+        Simulates the formation and evolution of contrails from a Flight
+        using the APCEMM plume model described in Fritz et. al. (2020)
+        :cite:`schumannRolePlumescaleProcesses2020`.
+
+        Parameters
+        ----------
+        source : Flight | None
+            Input Flight to model.
+        **params : Any
+            Overwrite model parameters before eval.
+
+        Returns
+        -------
+        Flight | NoReturn
+            Flight with exit status of APCEMM simulations. Detailed APCEMM outputs are attached
+            to model :attr:`vortex` and :attr:`contrail` attributes (see :class:`APCEMM` notes
+            for details).
+
+        References
+        ----------
+        - :cite:`fritzRolePlumescaleProcesses2020`
+        """
+
+        self.update_params(params)
+        self.set_source(source)
+        self.source = self.require_source_type(Flight)
+
+        logger.info("Attaching APCEMM initial conditions to source")
+        self.attach_apcemm_initial_conditions()
+
+        logger.info("Computing Lagrangian trajectories")
+        self.compute_lagrangian_trajectories()
+
+        # Select segments to run in APCEMM
+        # Defaults to all segments, but allows user to select a subset
+        segments = self.params["segments"]
+        if segments is None:
+            segments = sorted(self.source.dataframe.index.unique())
+
+        # Generate input files (serial)
+        logger.info("Generating APCEMM input files")  # serial
+        for segment in segments:
+            rundir = self.apcemm_file(segment)
+            if self.cachestore.exists(rundir) and not self.params["overwrite"]:
+                msg = f"APCEMM run directory already exists at {rundir}"
+                raise ValueError(msg)
+            if self.cachestore.exists(rundir) and self.params["overwrite"]:
+                shutil.rmtree(rundir)
+            self.generate_apcemm_input(segment)
+
+        # Run APCEMM (parallelizable)
+        logger.info("Running APCEMM")
+        self.run_apcemm(segments)
+
+        # Process output (serial)
+        logger.info("Postprocessing APCEMM output")
+        self.process_apcemm_output()
+
+        return self.source
+
+    def attach_apcemm_initial_conditions(self) -> None:
+        """Compute fields required for APCEMM initial conditions and attach to source.
+
+        This modifies :attr:`source` by attaching quantities derived from meterology
+        data and aircraft performance calculations.
+        """
+
+        self._attach_apcemm_time()
+        self._attach_initial_met()
+        self._attach_aircraft_performance()
+
+    def compute_lagrangian_trajectories(self) -> None:
+        """Calculate Lagrangian trajectories using a :class:`DryAdvection` model.
```

Comment:
I would add further description here about what these lagrangian trajectories represent, e.g. these are the expected advection paths of contrail segments formed by the input source.

The are limited by the `param["max_age"]` with a `dt_integration` of `params["dt_lagrangian"]`.

Reply:
1dc2173b5f20445833c83ac16d15a3571d3001d9